### PR TITLE
Adding refresh_token_request and access_token_request compliance hooks

### PR DESCRIPTION
**Pull request by violuke**
This is based upon #214, so thanks to @jimcortez for basically solving this for me :+1: :pray:

We needed this for a custom internal OAuth solution hence no included compliance fix example.

This solved the problem we're having and I believe will solve some or all of the points in issues #182, #379, #430 and #264 so should be useful for the wider community, despite not including a specific compliance fix example.

Here's our internal compliance fix:

```py
import base64
from oauthlib.common import add_params_to_uri

def internal_oauth_compliance_fix(session, client_secret: str):
    def _non_compliant_auth_header(url, headers, body):
        basic_auth_value = "{}:{}".format(session._client.client_id, client_secret)
        headers["Authorization"] = 'Basic {}'.format(base64.b64encode(basic_auth_value.encode("utf-8")).decode("utf-8"))
        token = [('token', session._client.access_token)]
        url = add_params_to_uri(url, token)
        return url, headers, body

    session.register_compliance_hook('refresh_token_request', _non_compliant_auth_header)
    return session
```

I hope this helps someone else, thanks again to @jimcortez for the original work and if this could be merged soon that would be much appreciated. :+1:

### CODE REVIEW

**Comment by timdawborn** [comment]
An example of a large public application that needs Basic Auth for the refresh tokens is the Xero API: https://developer.xero.com/documentation/oauth2/auth-flow#refresh

**Comment by violuke** [comment]
I've just also added support for an `access_token_request` hook. We were in the same situation as #244 (needing to send the data as json on a password grant type), but in our case the API was unable to accept via `application/x-www-form-urlencoded`. We also had to remove the `redirect_uri` key from the JSON or the request was rejected.

I know this is against the Oauth2 specs, but that's the whole point of compliant hooks :wink: Here's an example usage:

```python
def redacted_company_compliance_fix(session):
    def _request_in_json_not_form_urlencoded(url, headers, request_kwargs):
        headers["Content-Type"] = "application/json"

        if "redirect_uri" in request_kwargs["data"]:
            del request_kwargs["data"]["redirect_uri"]

        request_kwargs["json"] = request_kwargs["data"]

        del request_kwargs["data"]

        return url, headers, request_kwargs

    session.register_compliance_hook("access_token_request", _request_in_json_not_form_urlencoded)
    return session
```

**Comment by violuke** [comment]
@jtroussard would be please be possible for this to be reviewed? I think it will help in lots of incompliant situations. Thank you 🙏

File: requests_oauthlib/oauth2_session.py [hunk 1]
```diff
@@ -337,6 +338,10 @@ def fetch_token(
             request_kwargs["params"] = dict(urldecode(body))
         else:
             raise ValueError("The method kwarg must be POST or GET.")
+            
+        for hook in self.compliance_hook["access_token_request"]:
```
Comment (JonathanHuot) [inline requests_oauthlib/oauth2_session.py:342]: Is this access_token_request declared somewhere?

**Review by JonathanHuot — CHANGES_REQUESTED** [review]
Would it be possible to add at least a unit test? This feature is a great addition, so thanks for the extra work.

File: requests_oauthlib/oauth2_session.py (hunk 1 above)
Comment (violuke) [reply requests_oauthlib/oauth2_session.py:343]: Good point 🤦 I've corrected this.

I'm not sure the best way to go about a unit test for these as they don't change the response like the other compliance hooks. Any suggestions or advice would be appreciated. Thank you.

**Comment by violuke** [comment]
@JonathanHuot I've added a test now for both of the new hooks. Thanks.

**Comment by JonathanHuot** [comment]
Can you execute `tox -e black` and `tox -e py27` as they seem not passing in pipeline. Thanks in advance

**Comment by violuke** [comment]
Both are sorted now, thanks.

**Review by JonathanHuot — APPROVED** [review]

**Comment by JonathanHuot** [comment]
Thanks for this useful feature!

**Comment by violuke** [comment]
@JonathanHuot would you please be able to make a release on PyPI for this so that we can switch our package manager over to the main project again? Thank you.

**Comment by violuke** [comment]
Any chance of that PyPI release? 😉

**Comment by JonathanHuot** [comment]
Hi @violuke, I will prepare a release this month. Thanks for your patience.

**Pull request merged**